### PR TITLE
Add indexOf function

### PR DESCRIPTION
**Pull request by svozza**
As mentioned in #35.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -374,6 +374,21 @@
     return Nothing();
   });
 
+  //. indexOf :: a -> [a] - Maybe Number
```
Comment (davidchambers) [inline index.js]: Your second arrow is missing its `>`.

File: index.js [hunk 2]
```diff
@@ -374,6 +374,21 @@
...
+  //.
+  //. Returns Just the position of the first occurrence of an
+  //. element in a list, or Nothing if the element does not occur
+  //. in the list.
+  //.
+  //. ```javascript
+  //. > S.indexOf(3, [1, 2, 3, 4, 5])
+  //. Just(2)
```
Comment (davidchambers) [inline index.js]: When dealing with indexes it's clearer to use a list of strings in the example.

File: index.js [hunk 3]
```diff
@@ -374,6 +374,21 @@
...
+  //. ```
+  S.indexOf = R.curry(function(target, list) {
+    var idx = R.indexOf(target, list);
+    return idx === -1 ? Nothing() : Just(idx);
```
Comment (davidchambers) [inline index.js]: I prefer not to specify the arbitrarily chosen failure value:

```javascript
return idx >= 0 ? Just(idx) : Nothing();
```

**Comment by davidchambers** [comment]
This is great! Have you any interest in defining `S.lastIndexOf` as well?

**Comment by svozza** [comment]
Yeah, definitely. Will I add it to this PR or make a new one?

**Comment by davidchambers** [comment]
> Will I add it to this PR or make a new one?

That's up to you. I'm happy with either option. :)

**Comment by svozza** [comment]
Nice one. Got a better name for the `indexOfToMaybe` function I've added there? Not happy with it at all.

**Comment by svozza** [comment]
Did you get my Gitter message btw?

File: index.js [hunk 4]
```diff
@@ -374,6 +376,36 @@
...
+  //. in the list.
+  //.
+  //. ```javascript
+  //. > S.indexOf('c', ['a','b', 'c', 'd', 'e'])
+  //. Just(2)
+  //. ```
+  S.indexOf = R.curry(function(target, list) {
+    var idx = R.indexOf(target, list);
+    return indexOfToMaybe(idx);
+  });
```
Comment (davidchambers) [inline index.js]: If we :curry: `filter` and add `var _ = R.__;` we can write:

```javascript
S.indexOf = R.curry(R.compose(filter(R.gte(_, 0)), Just, R.indexOf));
```

This was we could avoid having to come up with a good name for `indexOfToMaybe`. :)

File: index.js [hunk 5]
```diff
@@ -374,6 +376,36 @@
...
   });
 
+  //. indexOf :: a -> [a] -> Maybe Number
+  //.
+  //. Returns Just the position of the first occurrence of an
+  //. element in a list, or Nothing if the element does not occur
+  //. in the list.
+  //.
+  //. ```javascript
+  //. > S.indexOf('c', ['a','b', 'c', 'd', 'e'])
```
Comment (davidchambers) [inline index.js]: We're missing a space before `'b'`.

File: index.js [hunk 6]
```diff
@@ -374,6 +376,36 @@
...
+  });
+
+  //. lastIndexOf :: a -> [a] -> Maybe Number
+  //.
+  //. Returns Just the position of the last occurrence of an
+  //. element in a list, or Nothing if the element does not occur
+  //. in the list.
+  //.
+  //. ```javascript
+  //. > S.indexOf('c', ['a','b', 'b', 'c', 'c'])
```
Comment (davidchambers) [inline index.js]: We're missing a space before `'b'` here as well.

File: test/index.js [hunk 7]
```diff
@@ -1213,6 +1213,40 @@ describe('list', function() {
...
   });
 
+  describe('indexOf', function() {
+
+    it('returns a Nothing for an empty list', function() {
+      assert(S.indexOf(10, []).equals(S.Nothing()));
+    });
+
+    it('returns a Nothing if the element is not found', function() {
+      assert(S.indexOf(10, [1, 2, 3, 4, 5]).equals(S.Nothing()), true);
```
Comment (davidchambers) [inline test/index.js]: This should either be `assert(...)` or `eq(..., true)`. :)

File: index.js (hunk 4 above)
Comment (svozza) [inline index.js]: Nice. But I don't think it gets us of the hook, this is a pattern that's going to be repeated in the `indexOf` methods for the strings so it'll still make sense to have a function that converts any function where a -1 indicates a `null` value and reuse it across the board.

**Comment by svozza** [comment]
So I wrote a wrapper function for indexOf methods, I'll add the string ones in a separate PR though.

File: index.js [hunk 8]
```diff
@@ -343,6 +343,10 @@
     return xBool !== yBool ? or(x, y) : xEmpty;
   });
 
+  var wrapIndexOf = function(f) {
+    return R.curry(R.compose(R.ifElse(R.lte(0), Just, Nothing), f));
```
Comment (benperez) [inline index.js]: This might be more clear with a placeholder:
```javascript
R.curry(R.compose(R.ifElse(R.gte(R.__, 0), Just, Nothing), f));
```

Comment (svozza) [inline index.js]: Yes, but would it be more beautiful? :smile: But yeah, I agree. Actually would there be a way to make the whole function definition point free?

Comment (davidchambers) [inline index.js]: I agree that the placeholder avoids misinterpretation of a partially applied `R.{lt,gt}{,e}`.

> Actually would there be a way to make the whole function definition point free?

Yesterday, between talks at LambdaConf, I did just that, but I didn't have a reliable internet connection so didn't post it.

```javascript
//  compose :: (b -> c) -> (a -> b) -> a -> c
var compose = R.curryN(2, R.compose);

var sanctifyIndexOf = R.pipe(compose(Just),
                             compose(filter(R.gte(_, 0))),
                             R.curry);
```

This is not necessarily as good as your version of the function, but it was an enjoyable exercise.

I do like the name `sanctifyIndexOf`, though. :)

Comment (svozza) [inline index.js]: Nice. Big +1 for `sanctifyIndexOf`. :)

**Comment by davidchambers** [comment]
:deciduous_tree:

**Pull request merged**